### PR TITLE
r-knitr: add v1.42

### PR DESCRIPTION
**Pull request by alecbcs**
Add r-knitr v1.42. 
 
**Test Plan:**
Built successfully using `gcc@10.4.0` on Debian 11.

### CODE REVIEW

**Review by tldahlgren — COMMENTED** [review]
Confirmed the sha256.  Deferring merge decision to @glennpj .

File: var/spack/repos/builtin/packages/r-knitr/package.py [hunk 1]
```diff
@@ -31,17 +32,19 @@ class RKnitr(RPackage):
     depends_on("r@3.1.0:", type=("build", "run"), when="@1.15:1.22")
     depends_on("r@3.2.3:", type=("build", "run"), when="@1.23:")
     depends_on("r@3.3.0:", type=("build", "run"), when="@1.39:")
+
     depends_on("r-evaluate@0.10:", type=("build", "run"))
     depends_on("r-evaluate@0.15:", type=("build", "run"), when="@1.39:")
     depends_on("r-highr", type=("build", "run"))
-    depends_on("r-stringr@0.6:", type=("build", "run"))
+    depends_on("r-stringr@0.6:", type=("build", "run"), when="@:1.40")
```
Comment (glennpj) [inline var/spack/repos/builtin/packages/r-knitr/package.py:39]: This is another nit-picky thing but can you move this no longer active dependency to the end of the file? That keeps the inactive dependencies grouped together.

**Review by glennpj — APPROVED** [review]

File: var/spack/repos/builtin/packages/r-knitr/package.py (hunk 1 above)
Comment (glennpj) [inline var/spack/repos/builtin/packages/r-knitr/package.py]: This is another nit-picky thing but can you move this no longer active dependency to the end of the file? That keeps the inactive dependencies grouped together.

**Review by glennpj — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/r-knitr/package.py (hunk 1 above)
Comment (alecbcs) [reply var/spack/repos/builtin/packages/r-knitr/package.py]: Done! Apologies on the delay. Also sorted the dependencies for the sake of my own OCD lol.

**Review by tldahlgren — APPROVED** [review]

**Review by glennpj — APPROVED** [review]

**Pull request merged**